### ISSUE TITLE
Positive semi-definite requirement

### ISSUE THREAD

**Issue by derekbeaton** [jennyrieck/C-MARINeR#15]
A very serious issue for us to resolve. covSTATIS (and PCA) require covariance matrices that are square, symmetric, and positive semi definite (PSD). The square/symmetric is easy and an enforced requirement. 

PSD, on the other hand, is not always true for some "connectivity" data. These matrices do not adhere to the definition of a covariance matrix, but do contain correlation/covariance values. These typically result from selective use of items to correlate, e.g., skipping over certain pairs but different pairs, for each of the pairwise cor/cov. 

This poses a theoretical/computational/programmatic problem. There are three options:

* Strictly enforce the cov part of covSTATIS; possibly at the loss of wide usability because of lack of adherence to theoretical assumptions/requirements of cov matrices & PCA.

* Only enforce square/symmetric, and issue warnings about non-PSD matrices; could be supplemented by diagnostic information so users can see which matrices are the offending ones

* Take in any square/symmetric sets of matrices and then internally enforce PSD via two possibilities: (1) relatively high tolerance for eigenvalues (a user controlled parameter), or (2) reconstruct every square/symmetric matrix with only vectors that are valid (and be a bit strict): with only vectors that have strictly positive eigenvalues (which now changes the condition to positive definite).

**Comment by derekbeaton** [comment]
The final point also brings up two possibilities: apply this to *each* matrix, or only the compromise? I suppose that, so long as the compromise matrix is PSD, this is good enough. However, the MFA norm poses a problem at the per-matrix level.

**Comment by derekbeaton** [comment]
These issues can also result from having more ROIs/voxels than time points.

**Comment by derekbeaton** [comment]
I think this could be handled in one of two ways for now: either in the GSVD package, I rewrite geigen() or tolerance.eigen() to use the SVD for stability, or, I set a much higher threshold like in nearPD (1e-6)

**Comment by derekbeaton** [comment]
I think I'll take the approach of using the svd() but making sure that the matrix is strictly square and symmetric, with a check for singular/eigenvalues with high negative magnitude and/or any complex values (but these shouldn't occur).

**Comment by derekbeaton** [comment]
I don't know anymore. Will re-visit this periodically to make a suitable decision. For now, there is a switch to just let offending matrices pass through

**Comment by derekbeaton** [comment]
After much hemming and hawing, and excessive indecision, I finally fixed this. It required two fixes.
One set of fixes was in the GSVD package, where I had to better/correctly handle the tolerance checks of eigen/singular values. The second set was to use tolerance via covstatis() to pass through to GSVD while also allowing us to strictly or not enforce the positive semi-definite requirement.

Phew.
